File: Java/word-break-ii.java
//  Word Break II

// Time O(N^2), Space O(N^2)
class Solution {
    public List<String> wordBreak(String s, List<String> wordDict) {
        int size = s.length();
        boolean[] f = new boolean[size+1];
        f[0] = true;
        boolean[][] words = new boolean[size][size];
        for (int i=1; i<=size; i++) {
            for (int j=0; j<i; j++) {
                if (f[j] && wordDict.contains(s.substring(j, i))) {
                    f[i] = true;
                    words[j][i-1] = true;
                }
            }
        }
        List<String> res = new ArrayList<String>();
        List<String> path = new ArrayList<String>();
        dfs(s, words, 0, path, res);
        return res;
    }
    
    private void dfs(String s, boolean[][] words, int cur, List<String> path, List<String> res) {
        if (cur == s.length()) {
            String temp = "";
            List<Integer> l = new ArrayList(path);
            for (int i=0; i<l.size(); i++) 
                temp = temp + l.get(i) + " ";
            res.add(temp);
        }
        for (int i=0; i<s.length(); i++) {
            if (words[cur][i]) {
                path.add(s.substring(cur, i+1));
                dfs(s, words, i+1, path, res);
                path.remove(path.size()-1);
            }
        }
    }
}

File: Leetcode150/word-break-ii.cc
//  Word Break II

class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        vector<bool> dp(s.length()+1, false);
        vector<vector<bool>> segment(s.length(), vector<bool>(s.length(), false));
        dp[0] = true;
        for (int i = 0; i < s.length(); i ++) {
            for (int j = 0; j <= i; j ++) {
                string temp = s.substr(j, i-j+1);
                if (dp[j] && find(wordDict.begin(), wordDict.end(), temp)!=wordDict.end()) {
                    dp[i+1] = true;
                    segment[j][i] = true;
                }
            }
        }
        vector<string> res;
        vector<string> comb;
        dfs(s, 0, segment, comb, res);
        return res;
    }
    
    void dfs(string &s, int cur, vector<vector<bool>> &segment, vector<string> &comb, vector<string> &res) {
        if (cur == s.length()) {
            string temp = "";
            for (auto &word : comb)
                temp += word + " ";
            temp.erase(temp.end()-1);
            res.push_back(temp);
        }
        for (int i = cur; i < s.length(); i ++) {
            if (segment[cur][i]) {
                comb.push_back(s.substr(cur, i-cur+1));
                dfs(s, i+1, segment, comb, res);
                comb.pop_back();
            }
        }
    }
};
